browsable_source/2.6/Windows/Scilab-2.6/man/nonlinear/int2d.man
.TH int2d G "April 1993" "Scilab Group" "Scilab Function"
.so ../sci.an
.SH NAME
int2d - definite 2D integral by quadrature and cubature method
.SH CALLING SEQUENCE
.nf
[I,err]=int2d(X,Y,f [,params])
.fi
.SH PARAMETERS
.TP 10
X 
: a 3 by N array containing the abscissae of the vertices
  of the N triangles.
.TP 10
Y 
: a 3 by N array containing the ordinates of the vertices
  of the N triangles.
.TP
f
: external (function or list or string) defining the integrand \fVf(u,v)\fR;
.TP
params
: real vector \fV[tol, iclose, maxtri, mevals, iflag]\fR. default
value is \fV[1.d-10, 1, 50, 4000, 1]\fR.
.RS
.TP
tol
:the desired bound on the error.  If \fViflag=0\fR,
\fVtol\fR is interpreted as a bound on the relative error;
if \fViflag=\fR1, the bound is on the absolute error.
.TP
iclose
:an integer parameter that determines the selection
of LQM0 or LQM1 methods.  If \fViclose=1\fR then LQM1 is used.
Any other value of \fViclose\fR causes LQM0 to be used.
LQM0 uses function values only at interior points of
the triangle.  LQM1 is usually more accurate than LQM0
but involves evaluating the integrand at more points
including some on the boundary of the triangle.  It
will usually be better to use LQM1 unless the integrand
has singularities on the boundary of the triangle.
.TP
maxtri
:the maximum number of triangles in the final triangulation of the
region
.TP
mevals
: the maximum number of function evaluations
to be allowed.  This number will be effective in limiting
the computation only if it is less than 94*\fVmaxtri\fR  when LQM1
is specified or 56*\fVmaxtri\fR when LQM0 is specified.
.TP
iflag
:
.RE
.TP
I
: the integral value
.TP
err
: the estimated error
.SH DESCRIPTION
\fVint2d\fR computes the two-dimensional integral of a
function \fVf\fR over a region consisting of \fVn\fR triangles.
A total error estimate is obtained and compared with a
tolerance - \fVtol\fR - that is provided as input to the subroutine.
The error tolerance is treated as either relative or absolute
depending on the input value of \fViflag\fR.  A 'Local Quadrature
Module' is applied to each input triangle and estimates of the
total integral and the total error are computed.  The local
quadrature module is either subroutine LQM0 or subroutine
LQM1 and the choice between them is determined by the
value of the input variable \fViclose\fR.

If the total error estimate exceeds the tolerance, the triangle
with the largest absolute error is divided into two triangles
by a median to its longest side.  The local quadrature module
is then applied to each of the subtriangles to obtain new
estimates of the integral and the error.  This process is
repeated until either (1) the error tolerance is satisfied,
(2) the number of triangles generated exceeds the input
parameter \fVmaxtri\fR, (3) the number of integrand evaluations
exceeds the input parameter \fVmevals\fR, or (4) the function
senses that roundoff error is beginning to contaminate
the result.
.SH EXAMPLES
.nf
X=[0,0;1,1;1,0];
Y=[0,0;0,1;1,1];
deff('z=f(x,y)','z=cos(x+y)')
[I,e]=int2d(X,Y,f)
// computes the integrand over the square [0 1]x[0 1]
.fi
.SH SEE ALSO
intc, intl, int3d, intg, mesh2d
.SH REFERENCES
Fortran routine twodq,Authors: Kahaner,D.K.,N.B.S., Rechard,O.W.,N.B.S.,
Barnhill,Robert,Univ. of UTAH


browsable_source/3.1.1/Windows/scilab-3.1.1/routines/interf/int2d.f
      subroutine int2d
C     --------------------------------------------
c     Scilab intg 
c      implicit undefined (a-z)
c     Copyright INRIA
      include '../stack.h'
      character*(5) fname
      character*(nlgh+1) namef
      integer iero 
      common/ierajf/iero
      external bint2d,fint2d
      double precision tol,result,erro
      logical getexternal,getrmat,type ,cremat,checkval
      integer topk,lr,katop,kydot,top2,lra,lrb,lc
c
      integer iadr,sadr
      external setfint2d
c
      iadr(l)=l+l-1
      sadr(l)=(l/2)+1
      fname='int2d'
      if(rhs.lt.3) then
         call error(39)
         return
      endif
      type=.false.
      top2=top
      topk=top
      if(rhs.eq.4) then
         if (.not.getrmat(fname,topk,top,m1,n1,lr))  return
         if(.not.checkval(fname,m1*n1,5)) return
         tol=max(stk(lr),0.0d0)
         iclose=stk(lr+1)
         maxtri=max(int(stk(lr+2)),1)
         mevals=max(int(stk(lr+3)),1)
         iflag=stk(lr+4)
         if((iclose.ne.0.and.iclose.ne.1) .or. 
     $        (iflag.ne.0.and.iflag.ne.1)) then
            err=4
            call error(36)
         endif
         top=top-1
      else
         tol=1.d-10
         iclose=1
         maxtri=50
         mevals=4000
         iflag=1
      endif
c     
      if (.not.getexternal(fname,topk,top,namef,type,setfint2d)) return
      kext=top
      top=top-1
      
      if (.not.getrmat(fname,topk,top,my,ny,ly))  return
      if(.not.checkval(fname,my,3)) return

      top=top-1
      kxtop=top

      if (.not.getrmat(fname,topk,top,mx,nx,lx))  return
      if(.not.checkval(fname,mx,3)) return


      if(.not.checkval(fname,nx,ny)) return

c     definition des variables pour l'external
      top=top2+1
      kxx=top
      if (.not.cremat(fname,top,0,1,1,lxx,lc)) return

      top=top+1
      kyy=top
      if (.not.cremat(fname,top,0,1,1,lyy,lc)) return

c     tableaux de travail 
      lw=9*maxtri
      top=top+1
      if (.not.cremat(fname,top,0,1,lw,ldata,lc)) return

c     tableau de travail entier necessaire 
      liw=2*maxtri
      top=top+1
      if (.not.cremat(fname,top,0,1,iadr(liw)+1,liwork,lc)) return

c
c     external scilab
c
      top=top+1
      ipal=iadr(lstk(top))
      istk(ipal)=1
      istk(ipal+1)=ipal+2
      istk(ipal+2)=kext
      istk(ipal+3)=kxx
      istk(ipal+4)=kyy
      lstk(top+1)=sadr(ipal+5)

      nu=0
      nd=0
      if(type) then 
         call TWODQ(fint2d,nx,stk(lx),stk(ly),tol,iclose,maxtri,mevals
     $        ,result,erro,nu,nd,nevals,iflag,stk(ldata),stk(liwork))

      else
         call TWODQ(bint2d,nx,stk(lx),stk(ly),tol,iclose,maxtri,mevals
     $        ,result,erro,nu,nd,nevals,iflag,stk(ldata),stk(liwork))

      endif

      if(err.gt.0)return
      if(iflag.gt.0) then
         if(iflag.eq.1) then
c     termination for lack of space to divide triangle
            call msgs(81,0)
         elseif(iflag.eq.2) then
c     termination because of roundoff noise
            call msgs(82,0)
         elseif(iflag.eq.3) then
c     termination for relative error <5.0*%eps
            call msgs(83,0)
         elseif(iflag.eq.4) then
c     termination: number of function evaluations > MEVALS
            call msgs(84,0)
         endif
      endif
      top=top2-rhs+1
      if (.not.cremat(fname,top,0,1,1,lr1,lc)) return
      stk(lr1)=result
      if(lhs.ge.2) then
         top=top+1
         if (.not.cremat(fname,top,0,1,1,lr2,lc)) return
         stk(lr2)=erro
      endif
      if(lhs.eq.3) then
         top=top+1
         if (.not.cremat(fname,top,0,1,1,lr2,lc)) return
         stk(lr2)=nevals
      endif
      return
      end

